DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,696 @@
+    immutable int* p0 = safePtr;   // allowed, @safe variable
+    immutable int* p1 = systemPtr; // error, reading @system var of unsafe type
+    int i0 = safeInt;              // allowed
+    int i1 = systemInt;            // allowed, not an unsafe type
+}
+```
+
+**(3) Variables and fields without annotation are `@safe` unless their initial value is not `@safe`**
+
+The rules regarding variables and fields are as follows:
+- An initialization expression `x` is `@system` when the function `(() => x)` is inferred as `@system`.
+- When marked `@system`, the result is always `@system` regardless of the type.
+- When marked `@trusted`, the initialization expression `x` is treated as `(() @trusted => x)`.
+- When marked `@safe`, the initialization expression must be `@safe`.
+- In the absence of an annotation, the result is `@system` only if the type is unsafe and the initialization expression is `@system`.
+
+```D
+int* getPtr() @system {return cast(int*) 0x8035FDF0;}
+int  getVal() @system {return -1;}
+
+extern int* x0;                   // @safe by default
+int* x1 = x0;                     // @safe, (() => x0) is @safe
+int* x2 = cast(int*) 0x8035FDF0;  // @system, (() => cast(int*) 0x8035FDF0) is @system
+int* x3 = getPtr();               // @system, (() => getPtr()) is @system
+int  x4 = getVal();               // @safe, int is not an unsafe type
+@system int x5 = 1;               // @system as requested
+@trusted int* x6 = getPtr();      // @safe, the getPtr call gets trusted
+@safe int* x7 = getPtr();         // error: cannot initialize @safe variable with @system initializer
+
+struct S {
+    // same rules for fields:
+    int* x9 = x3; // @system
+    int  x8 = x5; // @safe
+}
+```
+
+An exception to the last rules is made on unsafe types when the compiler knows the resulting value is safe.
+```D
+int* getNull() pure @system {return null;}
+int* n = getNull(); // despite unsafe type with @system initialization expression, inferred as @safe
+```
+
+Annotations with a scope (`@system {}`) or colon (`@system:`) affect variables just like they do functions.
+```D
+@system {
+    int y0; // @system
+}
+
+@system:
+int y1; // @system
+```
+
+**(4) `__traits(getFunctionAttributes)` may be called on variables and fields**
+
+Currently it is possible to give function attributes to declarations that aren't functions.
+It is not possible however to inspect any of them.
+
+```D
+@system @nogc pure nothrow int x;
+pragma(msg, __traits(getFunctionAttributes, x)); // error: first argument is not a function
+pragma(msg, __traits(getAttributes, x)); // tuple()
+```
+
+Since memory safety-related attributes now have an effect on variables and fields, it becomes useful to inspect them.
+Therefor the restriction on the getFunctionAttributes trait gets lifted.
+
+The name "function attributes" is a bit unfortunate in this case, but this DIP does not aim to fix that.
+
+**(5) `bool` becomes an unsafe type**
```

Comment:
Wouldn't it be better instead to make it `@system` to somehow construct a `bool` that is neither 0 or 1?

Reply:
That's exactly what I'm proposing. Ways to construct such a bool are with a union, void initialization, array/pointer casting, and by making `bool` an unsafe type those actions become forbidden in `@safe` code.

Reply:
I think I understand now. Would you mind editing it so that it's clear *why* bool has to be an unsafe type? Thanks!

Reply:
I think the first two lines already do that? This line:

> By making `bool` an unsafe type, it can be both fast and `@safe` since all tricks to make invalid booleans are now disabled in `@safe` code.

Seems synonymous to me with:

> Wouldn't it be better instead to make it @system to somehow construct a bool that is neither 0 or 1?

Assuming that:
invalid == not 0 or 1
disabled in `@safe` == only `@system`